src/unziptuple.rs

```diff
@@ -0,0 +1,58 @@
+/// Unzips an iterator over tuples into a tuple of containers.
+///
+/// ```
+/// use itertools::multiunzip;
+///
+/// let inputs = vec![(1, 2, 3), (4, 5, 6), (7, 8, 9)];
+///
+/// let (a, b, c): (Vec<_>, Vec<_>, Vec<_>) = multiunzip(inputs);
+///
+/// assert_eq!((a, b, c), (vec![1, 4, 7], vec![2, 5, 8], vec![3, 6, 9]));
+/// ```
+pub fn multiunzip<FromI, I>(i: I) -> FromI
+where
+    I: IntoIterator,
+    I::IntoIter: MultiUnzip<FromI>,
+{
+    i.into_iter().multiunzip()
+}
+
+/// An iterator that can be unzipped into multiple collections.
+///
+/// See [`.multiunzip()`](crate::Itertools::multiunzip) for more information.
+pub trait MultiUnzip<FromI>: Iterator {
+    /// Unzip this iterator into multiple collections.
+    fn multiunzip(self) -> FromI;
+}
+
+macro_rules! impl_unzip_iter {
+    ($($T:ident => $FromT:ident),*) => (
+        impl_unzip_iter!(@rec $($T => $FromT,)*);
+    );
+    (@rec) => ();
+    (@rec $__:ident => $___:ident, $($T:ident => $FromT:ident,)*) => (
+        #[allow(non_snake_case)]
+        impl<IT: Iterator<Item = ($($T,)*)>, $($T, $FromT: Default + Extend<$T>),* > MultiUnzip<($($FromT,)*)> for IT {
+            fn multiunzip(self) -> ($($FromT,)*) {
+                let mut res = ($($FromT::default(),)*);
+                let ($($FromT,)*) = &mut res;
+
+                // Still unstable #72631
+                // let (lower_bound, _) = self.size_hint();
+                // if lower_bound > 0 {
+                //     $($FromT.extend_reserve(lower_bound);)*
+                // }
```

Comment:
That part confused me as to why a local function is being used there to create the closure instead of just using a closure normally.

Though currently we can't really do that either anyways as `extend_one` is unstable.

Reply:
> That part confused me as to why a local function is being used there to create the closure instead of just using a closure normally.

Not sure, but I think it helps by avoiding some monomorphizations. However, `itertools` does not go this route so far, so I'd keep your normal closure.

Reply:
Oh I just realized this was linking to something entirely different than I thought(I missed you saying old version so I assumed it linked to the current impl) so my reply didn't really make sense.

So ye we could consider doing that old trick to reserve some size for this if we actually wanted to I suppose.